cardano-api/src/Cardano/Api/LedgerState.hs

```diff
@@ -388,36 +367,212 @@ foldBlocks nodeConfigFilePath cardanoModeParams socketPath enableValidation stat
           clientIdle_DoneN
             :: Nat n -- Number of requests inflight.
             -> Maybe Text -- Return value (maybe an error)
-            -> IO (ClientPipelinedStIdle n (BlockInMode CardanoMode) ChainPoint ChainTip IO ())
+            -> IO (CSP.ClientPipelinedStIdle n (BlockInMode CardanoMode) ChainPoint ChainTip IO ())
           clientIdle_DoneN n errorMay = case n of
-            Succ predN -> return (CollectResponse Nothing (clientNext_DoneN predN errorMay)) -- Ignore remaining message responses
+            Succ predN -> return (CSP.CollectResponse Nothing (clientNext_DoneN predN errorMay)) -- Ignore remaining message responses
             Zero -> do
               writeIORef errorIORef errorMay
-              return (SendMsgDone ())
+              return (CSP.SendMsgDone ())
 
           clientNext_DoneN
             :: Nat n -- Number of requests inflight.
             -> Maybe Text -- Return value (maybe an error)
-            -> ClientStNext n (BlockInMode CardanoMode) ChainPoint ChainTip IO ()
+            -> CSP.ClientStNext n (BlockInMode CardanoMode) ChainPoint ChainTip IO ()
           clientNext_DoneN n errorMay =
-            ClientStNext {
-                recvMsgRollForward = \_ _ -> clientIdle_DoneN n errorMay
-              , recvMsgRollBackward = \_ _ -> clientIdle_DoneN n errorMay
+            CSP.ClientStNext {
+                CSP.recvMsgRollForward = \_ _ -> clientIdle_DoneN n errorMay
+              , CSP.recvMsgRollBackward = \_ _ -> clientIdle_DoneN n errorMay
               }
 
           fromChainTip :: ChainTip -> WithOrigin BlockNo
           fromChainTip ct = case ct of
             ChainTipAtGenesis -> Origin
             ChainTip _ _ bno -> At bno
 
+-- | Wrap a 'ChainSyncClient' with logic that tracks the ledger state.
+chainSyncClientWithLedgerState
+  :: forall m a.
+     Monad m
+  => Env
+  -> LedgerState
+  -- ^ Initial ledger state
+  -> Bool
```

Comment:
Can we use a custom sum type instead of a `Bool` here? Maybe `ValidationMode` or something like that would be suitable.

Reply:
Done!